### PR TITLE
[mlir][spirv] Add a generic `convert-to-spirv` pass

### PR DESCRIPTION
**Pull request by obtuseangleAZ**
This PR implements a MVP version of an MLIR lowering pipeline to SPIR-V. The goal of adding this pipeline is to have a better test coverage of SPIR-V compilation upstream, and enable writing simple kernels by hand. The dialects supported in this version include `arith`, `vector` (only 1-D vectors with size 2,3,4,8 or 16), `scf`, `ub`, `index`, `func` and `math`. New test cases for the pass are also included in this PR.

**Relevant links**

- [Open MLIR Meeting - YouTube Video](https://www.youtube.com/watch?v=csWPOQfgLMo)
- [Discussion on LLVM Forum](https://discourse.llvm.org/t/open-mlir-meeting-12-14-2023-discussion-on-improving-handling-of-unit-dimensions-in-the-vector-dialect/75683)

**Future plans**

- Add conversion patterns for other dialects, e.g. `gpu`, `tensor`, etc.
- Include vector transformation to unroll vectors to 1-D, and handle those with unsupported sizes.
- Implement multiple-return. SPIR-V does not support multiple return values since a `spirv.func` can only return zero or one values. It might be possible to wrap the return values in a `spirv.struct`.
- Add a conversion for `scf.parallel`.

### CODE REVIEW

File: mlir/include/mlir/Conversion/Passes.td
Comment (Hardcode84) [inline mlir/include/mlir/Conversion/Passes.td:43]: I think, current preferred style is without explicit `let constructor =` so tablegen will generate the constructor function.

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (Hardcode84) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:76]: nit: `applyPartialConversion(op, *target, std::move(patterns))`

Also, drop parents on single-line if

Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp]: Changed

File: mlir/include/mlir/Conversion/Passes.td
Comment (angelz913) [reply mlir/include/mlir/Conversion/Passes.td]: Removed the explicit `let constructor =`

File: mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h
Comment (kuhar) [inline mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h:7]: This header seems misaligned.

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:7]: Also here

File: mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h
Comment (kuhar) [inline mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h:12]: Do we still need it after dropping the custom constructor?

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:40]:
```suggestion
struct ConvertToSPIRVPass final
    : impl::ConvertToSPIRVPassBase<ConvertToSPIRVPass> {
```

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:47]: nit: use the full type when the type is not obvious based on the RHS only

File: mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h
Comment (angelz913) [reply mlir/include/mlir/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.h:7]: Fixed

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp]: Changed

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (angelz913) [reply mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp]: Fixed

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:43]:
```suggestion
```

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:44]:
```suggestion
  void runOnOperation() override {
```

File: mlir/test/Conversion/ConvertToSPIRV/arith.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/arith.mlir:40]: There's no std dialect anymore -- these are arith ops.

File: mlir/test/Conversion/ConvertToSPIRV/arith.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/arith.mlir:140]: I don't think we need to test all of these here -- one should be enough

File: mlir/test/Conversion/ConvertToSPIRV/arith.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/arith.mlir:180]: same here

File: mlir/test/Conversion/ConvertToSPIRV/arith.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/arith.mlir:205]: same here

File: mlir/test/Conversion/ConvertToSPIRV/index.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/index.mlir:63]: same here

File: mlir/test/Conversion/ConvertToSPIRV/index.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/index.mlir:85]: The exact expansion is tested in index to spirv. Here's I'd only check that the conversion succeeded and produced `spirv.ReturnValue`

File: mlir/test/Conversion/ConvertToSPIRV/index.mlir
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/index.mlir:100]: Also here

**Review by kuhar — APPROVED** [review]
LGTM but please wait for at least one more approval before merging.

File: mlir/test/Conversion/ConvertToSPIRV/ub.mlir
Comment (Hardcode84) [inline mlir/test/Conversion/ConvertToSPIRV/ub.mlir:5]: `[[UNDEF0:]]`

**Review by Hardcode84 — APPROVED** [review]

File: mlir/include/mlir/Conversion/Passes.td
Comment (kuhar) [reply mlir/include/mlir/Conversion/Passes.td:41]: @angelz913 talked about this in https://youtu.be/-qoMMrlYvGs?t=436 (starts around the 7m 15s mark). The TL;DR is we didn't know what interfaces to have, and decided to with a monolithic multi-stage v0 implementation, with the plan to add make it interface-based when gain confidence in this design.

Comment (joker-eph) [reply mlir/include/mlir/Conversion/Passes.td:41]: > The TL;DR is we didn't know what interfaces to have

I don't quite understand? Why isn't it just copy the convert-to-llvm one and rename it?

I have strong concerns with in-tree monolithic passes like this: what is the timeline to remove this and migrate to a pluggable one?

Comment (MaheshRavishankar) [reply mlir/include/mlir/Conversion/Passes.td:41]: I agree that in-tree monolithic passes arent great, but the change to use interfaces and go to more LLVM based approach is more involved (I dont think it is as simple as "copy the convert-to-llvm and rename it". Angel and Jakub can fill in more details here). I think timeline will depend on how much community involvement we get here. Jakub and Angel are trying to get upstream support flushed out more. So this is a strict improvement anyway.  Its probably better to iterate on this in a bit.

FWIW, for conversion to LLVM in IREE we just throw all the conversion patterns into one pass and run them together. I personally find the split of conversion from each dialect to LLVM kind of artificial. Everything needs to be translated to LLVM. Running multiple passes that walk the IR multiple times seems like a waste. But that is a downstream decision and not having monoliths in upstream MLIR is useful.

Comment (joker-eph) [reply mlir/include/mlir/Conversion/Passes.td:41]: >  I think timeline will depend on how much community involvement we get here. 

In this case I would want to see this pass moved to the test folder: I'm not comfortable with monolithic passes alongside the rest of the transformations right now.

> FWIW, for conversion to LLVM in IREE we just throw all the conversion patterns into one pass and run them together. I personally find the split of conversion from each dialect to LLVM kind of artificial

I suspect you missed the point: these are all made that way upstream for testing, the intention has always been that downstream project create a monolithic pass for their own purpose: that's exactly "work as intended".

Comment (MaheshRavishankar) [reply mlir/include/mlir/Conversion/Passes.td:41]: > > I think timeline will depend on how much community involvement we get here.
> 
> In this case I would want to see this pass moved to the test folder: I'm not comfortable with monolithic passes alongside the rest of the transformations right now.

I think this pass is used by the runner. Can we have a test pass that is used by the vulkan runner?  Anyway, I think there is agreement that this is the expected end-state, and the issue is noted.

Separate but related: We need to spell out clearly what passes are for in MLIR. I personally dont care about passes in MLIR. They are all only for testing IMO, but that is not a view held by everyone. If we explicitly call out that passes in MLIR are only for testing, unless really it needs to be a pass (like canonicalization, CSE, DCE or whatever).

Comment (joker-eph) [reply mlir/include/mlir/Conversion/Passes.td:41]: > I think this pass is used by the runner. Can we have a test pass that is used by the vulkan runner? 


The runner is supposed to be removed as far as I know (it seems that's another topic that needs a new push by the way...).

So tTat has me even more concerned now! This is a major misalignment to add new stuff to this runner when we agreed before that it should be removed...

Comment (kuhar) [reply mlir/include/mlir/Conversion/Passes.td:41]: The tool/binary should be retired and removed, by we want the functionality to be moved over to the (unified) mlir runner. Of course we don't want to just copy paste the code, but the dependency on the spirv conversion code will have to move over there in some form.

What you see across this is and following is incremental improvements towards having a cleaner lowering story for spirv upstream, but this is not the end state. Before the convert to spirv pass, we relied on the gpu to spirv pass which just added a bunch of random patterns to make the vulkan runner work. We are untangling all of this piece by piece.

Comment (MaheshRavishankar) [reply mlir/include/mlir/Conversion/Passes.td:41]: > > I think this pass is used by the runner. Can we have a test pass that is used by the vulkan runner?
> 
> The runner is supposed to be removed as far as I know (it seems that's another topic that needs a new push by the way...).
> 
> So tTat has me even more concerned now! This is a major misalignment to add new stuff to this runner when we agreed before that it should be removed...

Well, the runner is a tangent. W.R.T this pass itself, I think the end-state is what you are looking for. So it might be too monolithic to start with, but I think it is better to have it in tree an iterate on. In terms of moving it to a test pass, that is an option too.

W.R.T the runner, my only concern is that we have pretty woeful testing in MLIR, and without having some limited e2e runners that provide some basic gaurantees, Already I feel this testing hole is being filled by downstream projects like IREE (for example, almost every dialect conversion change that is being landed in MLIR is failing downstream, and then its a PROCESS to triage and provided repro and get through the part of is this a downstream issue or actual bug. I appreciate all the changes being made, but this is making it clear that there is just not enough testing in tree to catch these issues before hand. So we can drop the runner, but it will just make the situation worse).  So I think dropping the runner, or moving it to post-submit and then moving this pass to a test pass is an option, but requires someone to "own" all of this. I would rather let the stake-holders here (which for now is Lei and Jakub) decide how they want to manage the conversion to spir-v passes since they are also the users of this stuff downstream. They seem to agree that the end state is to be closer to LLVM lowering, but might get there in a few steps.

Comment (joker-eph) [reply mlir/include/mlir/Conversion/Passes.td:41]: >  but the dependency on the spirv conversion code will have to move over there in some form.

Why would it have to be moved to the runner? The current pattern is that the transformations are made with `mlir-opt` ahead of invoking the runner (hence why we ended up with a "pure" `mlir-cpu-runner`).

> So we can drop the runner, but it will just make the situation worse).

I think we're talking about different things maybe?
You seem to argue about the need for running some of these tests, while I'm just describing the system architecture of the project, which does not impact in any way which tests we're running.

That is, from a very high-level, the specific runner tool goes away and you do instead something like `mlir-opt -pass-pipeline="builtin.module(convert-to-spirv)" %s | mlir-cpu-runner --shared-libs=mlir_vulkan_runtime.so`.
(this is a transition we made for the `mlir-cuda-runner` ~3 years ago, and the Vulkan runner has been a TODO ever since IIUC)

Note that in this model, the test pass is available for the opt tool, the runner does not need to know about the test passes (or any pass).

Comment (kuhar) [reply mlir/include/mlir/Conversion/Passes.td:41]: Thanks for the clarification, Mehdi.

> You seem to argue about the need for running some of these tests, while I'm just describing the system architecture of the project, which does not impact in any way which tests we're running.

Right, that's a good way to put it. From my point of view, what I mostly care about is that we do keep these tests (as in, both the implementation of convert-to-spirv and the e2e .mlir tests) while the cleanup in this area (both gpu-to-spirv and the runner implementation). I think we all aggrege on the end state, so I just want to make sure that we can find a path that allows us to make incremental improvements to get there.

From your perspective, would it be an option to make this a test pass and **temporarily** have it registered it in the vulkan runner (similar to how they are manually listed in `mlir-opt.cpp`)? Maybe that's a middle ground, although I'm not sure of how much difference to a pass being a test or not there practically is for other users of mlir.

Comment (joker-eph) [reply mlir/include/mlir/Conversion/Passes.td:41]: > From your perspective, would it be an option to make this a test pass and temporarily have it registered it in the vulkan runner (similar to how they are manually listed in mlir-opt.cpp)?

The problem is that we optionally compile the test passes I think? (With `-DMLIR_INCLUDE_TESTS=ON`).

It may be easier to just keep the pass as-is until we migrate tests from mlir-vulkan-runner to mlir-cpu-runner?

The only thing that makes be nervous is that unless there is a timeline with people making progress on it, we'll still be adding mlir-vulkan-runner based tests multiple years from now. What I'm trying to see is some sort of a "gradient" on the progression of all this (and some timeline), rather than a quick immediate solution.

Comment (kuhar) [reply mlir/include/mlir/Conversion/Passes.td:41]: OK, let me check internally if we can commit to something more concrete along this axis. Overall, I think we did make a lot of progress on the spirv conversion test coverage and general cleanup in this area, but I understand why you'd want to prioritize the runner cleanup next.

Comment (kuhar) [reply mlir/include/mlir/Conversion/Passes.td:41]: @joker-eph I checked and we will have @andfau-amd work on the mlir runner migration, starting from ~next week.

**Pull request merged**
